### ISSUE TITLE
[feature request] Rename `Subset` -> `Resample` to reflect wider use

### ISSUE THREAD

**Issue by markalexander** [pytorch/pytorch#9559]
Very minor but I believe `torch.utils.data.dataset.Subset` might better be named `Resample` (or something similar), since it's possible to use for oversampling via duplicated indices.  This is a common use case for imbalanced data, and I had actually made my own before realizing that `Subset` was exactly the same indices-array method as I had done.

The switch would make it clear (a) that it can be used like this, and (b) that it *is* being used like this so that it doesn't get broken in some future update.

Thanks.

**Comment by weiyangfb** [comment]
Just want to comment on this, [Subset](https://github.com/pytorch/pytorch/blob/ba062e7da986bae28fea41f45bc85a1f8a74573d/torch/utils/data/dataset.py#L90-L106) looks more like an iterator to get a subset of data defined by `indices`. It is probably more appropriate to put Samplers under `torch.utils.data.sampler`

**Comment by markalexander** [comment]
@weiyangfb Agree.  In that vein, things like `SubsetRandomSampler` also have this extended use where the result is not necessarily a subsample.  Perhaps then `RandomIndicesSampler` is more appropriate, with a slight change to the docstring about replacement.  In fact `SubsetRandomSampler` is seemingly quite close to `torch.utils.data.dataset.Subset` in intent anyway, only it provides indices in random order.
